### ISSUE TITLE
Add feature encoding algorithms (Fisher vector, VLAD, and BoW) 

### ISSUE THREAD

**Issue by DavidTorpey** [scikit-image/scikit-image#4057]
I was wondering if it would be possible to include the three main popular feature encoding / pooling / quantisation algorithms into scikit-image, namely, Fisher vectors, VLAD, and BoW. They are very popular, time-tested algorithms with many applications.

I am happy to submit this feature as well.

**Comment by stefanv** [comment]
VLAD would complement DAISY well, I think. Are the other features applicable to images? You mention scikit-learn, but this is scikit-image.

**Comment by grlee77** [comment]
I am not very familiar with this area, but are there any patent / IP concerns that could make it difficult to incorporate these into scikit-image?  I believe our policy is not to distribute any patented algorithms so users don't have to worry about IP issues.

We recently removed seam carving for this reason and I thought the reason we didn't have SIFT was also due to patent issues. (Although for SIFT it looks like the patent is from 1999, so it may be expired?).

**Comment by stefanv** [comment]
For SIFT, we have Censure, which is arguably better. But if the patent expired we could look into it.

**Comment by DavidTorpey** [comment]
No, I don't think there are IP or patents concerns.

SIFT is separate to these encoding techniques. These encoding techniques such as VLAD and Fisher vectors simply encode a set of descriptors. These descriptors are left to the user to define, and can be anything from SIFT, SURF, ORB, KAZE, or any other descriptor at all.

**Comment by DavidTorpey** [comment]
The SIFT patent expires in March next year as far as I'm aware. But again, that does not affect the implementation and/or inclusion of the above algorithms in scikit-image.

**Comment by stefanv** [comment]
KAZE looks neat; we should try and get that in as well, if it is not patent encumbered.

**Comment by stefanv** [comment]
Wrt feature encoding, perhaps scikit-learn is indeed the more obvious home. Have you asked them? If they can't or won't, we can think about how to fit it into scikit-image.

**Comment by DavidTorpey** [comment]
Yeah, I don't think KAZE is patented.

Yeah, I've asked them. They would rather not include it there since it doesn't fit into the pipelines paradigm?

I would be happy to submit these algos.

**Comment by stefanv** [comment]
Go for it. We'll need a bit of justification, probably in the form of a gallery example, but otherwise I think it should be fine.

**Comment by DavidTorpey** [comment]
Okay perfect, no problem.

**Comment by DavidTorpey** [comment]
Can I bring in scikit-learn as a dependency, for learning the visual codebooks (Gaussian Mixture Models, KMeans clusterings, etc.)?

**Comment by stefanv** [comment]
scikit-learn is only a documentation dependency, and too heavy to add just for one submodule. This, again, is why the functionality is probably a better fit for that library.

**Comment by DavidTorpey** [comment]
Makes sense. I will have a chat with the scikit-learn guys. If they won't budge, I'll implement the codebook learning using numpy?

**Comment by hmaarrfk** [comment]
> scikit-learn is only a documentation dependency, and too heavy to add just for one submodule. This, again, is why the functionality is probably a better fit for that library.

I mean, we could just throw a friendly error message if we can't import sklearn.

Because scikit-image doesn't import the whole namespace at once, modular dependencies are possible.

**Comment by stefanv** [comment]
True; I guess we already have optional dependencies like dask.
